rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
+the amount of information provided. And the kernels should also demonstrate
+reasonably good performance otherwise, the creation overhead will be replaced by
+a suboptimal performance overhead, which will defeat the purpose of having the
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
+be global similar to the primitive cache.
+
+Pros:
+* Doesn't require users to modify their code to make the feature work
+* Eliminates the kernel creation overhead completely when it's possible
+* Doesn't require to expose kernel implementation details. For example, for
+the runtime dimensions feature we will have to document what dimensions can
+be runtime without introducing performance regressions
+
+Cons:
+* Primitive and primitive descriptor creation overhead will stay
+
+#### Key and Value
+
+The cache implementation will reside in the `common` part of the library. In order
+to abstract away `key` and `value` entities we will need to introduce the
+following abstractions:
+
+```cpp
+namespace kernel_cache {
+// Key
+struct key_t {
+    key_t(key_impl_t *impl);
+    virtual ~key_t() = default;
+
+    bool operator==(const key_t &other) const;
+    size_t hash() const;
+private:
+    std::shared_ptr<key_impl_t> impl_;
+};
+
+// The `key_impl_t` class will be the base class for all `key` entities.
+// Each engine kind and vendor will be able to define it the way the need.
+struct key_impl_t {
+    key_impl_t() = default;
+    virtual ~key_impl_t() = default;
+
+    key_impl_t(const key_impl_t &) = delete;
+    key_impl_t &operator=(const key_impl_t &) = delete;
+
+    virtual bool compare(const key_impl_t *key_impl) const = 0;
+    virtual size_t hash() const = 0;
+};
```

Comment:
I agree, runtime polymorphism is necessary, and it is fundamental in what I proposed. The issue that I have is the kernel cache system is one where there are many different key/objects types with ill-defined and commonly changing interfaces. In addition, the cache structure is fault tolerant to many failure modes of the key/object pairs. This creates a system where it is **very** difficult to certify oneDNN is behaving as intended, especially if we limit testing to exposed API interfaces as we have traditionally done. Because of this, consideration for common tools and requirements to prevent/catch programming errors seems warranted.

As one example, the use of uninitialized memory in hashing/equality comparison, but not object creation, is a fault tolerant failure we are extremely likely to encounter (and is an issue we already hit multiple times using cl_cache in gpu code). With the technique I mentioned, it is possible to implement something like the following:
```C++
template <typename T>
struct zero_initalized {
       zero_initalized() {
           std::memset(static_cast<uint8_t *>(this), 0, sizeof(T));
       }
}

struct my_key_impl_t: public zero_initalized<my_key_impl_t> {
       ... normal implementation...
}

```

and then require, via a templated constructor, that the `key_t` only contains `my_impl_t` which are `zero_intialized<my_impl_t>`. By doing this, we effectively eliminate the possibility of an uninitialized integer in the key implementation causing unnoticed issues. As far as I can see, it is not possible to make such global guarantees with the current design.

Reply:
@rjoursler could you clarify what kind of checks you want to make? 

I understand the point of making key comparison robust, but I am not sure if the number of points of failures will be reduced. e.g. if we want to check that all elements of the key are hashed/compared, we could have a common hash/compare function in based class and:
- either a different derived key type for each combination of fields required for a key. This could lead to a lot of derived classes. Could be simplified with templatization as you mentioned, but each instantiation point in the code becomes a point of failure
- or to have a common key type, and implementations will need to register which field of the key will be used. Same as above, registration becomes a point of failure.

With the current proposal, each derived class implementation becomes a point of failure. My point here is: the number of points of failures and boilerplate code depends on how many impls will share the same derived key. So it kind of makes sense to leave that part of the design not be in common but in backend implementations.

Reply:
@mgouicem, there are some common case dependent strategies that can be deployed. The main example that comes to mind from [my PR](https://github.com/intel-innersource/libraries.performance.math.onednn/pull/4946) is the following pattern:
```C++
// Can only be used with structures T meeting certain requirements
template <typename T>
struct hashable {
     // Zero initializes structure
     hashable() { std::memset(reinterpret_cast<uint8_t *>(this), 0, sizeof(T)); }
     // Prevents loss of zero initalization
     hashable(const hashable & other) { std::memmove(this, &other, sizeof(T)}
    // More code to prevent loss of zero initialization
     ...
    // Zero initialization makes this well defined
    bool operator==(const T& other) const { std::memcmp(this, &other, sizeof(T)}
    ...
    // Zero initialization makes this well defined
    hash() { return hash_range(reinterpret_cast<uint8_t *>(this), sizeof(T); }
    ...
}
// std::enable_if ensures hashable is valid except for a couple of failure modes
template <typename T, typename = std::enable_if<std::is_trivial<T>::value>>
trivial_kernel_impl_t : public kernel_impl_t { 
   bool compare( const key_impl_t * key_impl) const override {
        if(typeof(this) == typeof(key_impl) {
            return value == (static_cast<trivial_impl_t *> key_impl)->value); 
        }
        return false;
  }
   size_t hash() const override {
         return T.hash();
   }
private:
   hashable<T> value;
}
```
The only failure mode for `trivial_kernel_impl_t` is to "add a runtime pointer/reference to the structure" or to "sneak a non-zeroinitialized T into the structure". A few variants of this pattern should be usable for most kernel keys as, if the structure needs to be more complicated, caching it will lead to a bad time. Identifying these patterns and consolidating them in `common` will help prevent failures.

The other pattern that comes to mind is to require an interface like `virtual bool supports(engine_t * engine) = 0` to `key_impl_t` so that each implementation is forced to consider what `engine_t` structures are compatible with the key. By making this explicitly required, we help prevent any `multi-engine` scenarios being forgotten.